front/pages/api/w/[wId]/subscriptions/stripe_subscription_valid.ts

```diff
@@ -0,0 +1,188 @@
+import { getStripeSubscription } from "@app/lib/plans/stripe";
+import { apiError, withLogging } from "@app/logger/withlogging";
+
+// TODO(2024-04-05,pr): immediately after flav's merge, use the global constant
+const REPORT_USAGE_VALUES = ["MAU_1", "MAU_5", "MAU_10", "PER_SEAT"];
+
+export type GetSubscriptionResponseBody =
+  | { valid: true }
+  | { valid: false; reason: string };
+
+async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<WithAPIErrorReponse<GetSubscriptionResponseBody>>
+): Promise<void> {
+  const session = await getSession(req, res);
+  const auth = await Authenticator.fromSession(
+    session,
+    req.query.wId as string
+  );
+
+  const stripeSubscriptionId = req.query.stripeSubscriptionId as string;
+  if (!stripeSubscriptionId || typeof stripeSubscriptionId !== "string") {
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "invalid_request_error",
+        message: "Invalid request, missing stripeSubscriptionId.",
+      },
+    });
+  }
+
+  const owner = auth.workspace();
+  if (!owner) {
+    return apiError(req, res, {
+      status_code: 404,
+      api_error: {
+        type: "workspace_not_found",
+        message: "The workspace was not found.",
+      },
+    });
+  }
+
+  // Callers will be either superUsers (from poke) or admins of the workspace (from webhook calls)
+  if (!auth.isAdmin()) {
+    return apiError(req, res, {
+      status_code: 403,
+      api_error: {
+        type: "workspace_auth_error",
+        message:
+          "Only users that are `admins` for the current workspace can access this endpoint.",
+      },
+    });
+  }
+
+  // for admins of the workspace (webhook call), we must already have the
+  // subscription in our DB (matching the workspace)
+  const subscription = auth.subscription();
+  if (
+    !auth.isDustSuperUser() &&
+    (!subscription ||
+      subscription.stripeSubscriptionId !== stripeSubscriptionId)
+  ) {
+    return apiError(req, res, {
+      status_code: 404,
+      api_error: {
+        type: "subscription_not_found",
+        message: "The subscription was not found.",
+      },
+    });
+  }
+
+  let stripeSubscription: Stripe.Subscription;
+  try {
+    stripeSubscription = await getStripeSubscription(stripeSubscriptionId);
+  } catch (error) {
+    return apiError(req, res, {
+      status_code: 500,
+      api_error: {
+        type: "internal_server_error",
+        message: `Error while fetching subscription from stripe.\n${error}`,
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "GET": {
+      // Various validity checks for the subscription
+
+      // for superUsers (poke call), we check that the subscription is either
+      // not yet attributed to a workspace, or is attributed to the auth's
+      // workspace
+      if (
+        auth.isDustSuperUser() &&
+        subscription &&
+        subscription.stripeSubscriptionId !== stripeSubscriptionId
```

Comment:
So if I understand correctly, once a workspace has a subscription attached, we can't change it from Poke, is that correct?

Reply:
You can attach a new one from poke. What you can't do (and what is checked here) is attach a subscription that is already attached to another workspace, so when in poke you check if a subscription is valid for a workspace, it will answer 'no' if the subscription is attached to another workspace

Reply:
Are we sure about this? I don't see where we fetch the subscription in our DB from the stripe subscription 🤔 .

Reply:
As per IRL, you were correct; the check is not always relevant and not "validity" per se, so it's only in the poke route (following refactoring discussed below) and optional